### PR TITLE
Implement fermionic PureFock particle number measurement

### LINKED ISSUE #522: Implement `ParticleNumberMeasurement` for `fermionic.PureFockSimulator`
### What should be added?

The following should work:
```
import piquasso as pq

from scipy.stats import unitary_group

with pq.Program() as program:
    pq.Q() | pq.StateVector([1, 1, 1, 0, 0])
    pq.Q(all) | pq.Interferometer(unitary_group.rvs(5))
    pq.Q() | pq.ParticleNumberMeasurement()

simulator = pq.fermionic.PureFockSimulator(d=5)
result = simulator.execute(program, shots=1000)
result.samples  # -> [(0, 1, 1, 0, 1), (1, 1, 0, 0, 1), ...]
```

### PR DESCRIPTION
**Pull request by Rajesh270712**
Closes #522.

## Summary

- add `ParticleNumberMeasurement` support for `pq.fermionic.PureFockSimulator`
- return projected branch states for exact `shots=None` measurements
- add focused fermionic pure-Fock measurement coverage

## Validation

- `python -m pytest -q tests/fermionic/fock/test_measurements.py` (`3 passed`)
- `python -m pytest -q tests/fermionic/fock/test_state.py tests/fermionic/fock/test_measurements.py tests/fermionic/test_simulator_equivalence.py` (`43 passed`)
- `python -m pytest -q tests/fermionic/fock` (`58 passed`, 1 existing warning)
- `git diff --check upstream/main...HEAD`

## AI disclosure

I used OpenAI Codex to help inspect the codebase and draft the patch. I reviewed the changes and used the validation above before opening this PR.

### CODE REVIEW

File: tests/fermionic/fock/test_measurements.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+
+import piquasso as pq
+
+
+def test_ParticleNumberMeasurement_shots_None_projects_remaining_modes():
```
Comment (Kolarovszki) [inline tests/fermionic/fock/test_measurements.py:21]: Can you implement another test with shots not equal to None?

File: piquasso/fermionic/fock/simulation_steps.py [hunk 2]
```diff
@@ -103,6 +110,98 @@ def state_vector(
...
+        connector=connector,
+        config=config_copy,
+    )
+
+    remaining_modes = get_auxiliary_modes(state.d, modes)
+    fock_space_basis = get_fock_space_basis(state.d, state._config.cutoff)
+    measured_modes = list(modes)
+
+    for index, occupation_number in enumerate(fock_space_basis):
+        if tuple(occupation_number[measured_modes]) != subspace_basis:
```
Comment (Kolarovszki) [inline piquasso/fermionic/fock/simulation_steps.py]: I don't think this is the most efficient way to do this. Can you write a smaller loop and avoid this conditional?

**Review by Kolarovszki — CHANGES_REQUESTED** [review]

File: tests/fermionic/fock/test_measurements.py (hunk 1 above)
Comment (Rajesh270712) [reply tests/fermionic/fock/test_measurements.py:21]: Added `test_ParticleNumberMeasurement_finite_shots_projects_remaining_modes` for `shots=3`; it verifies the finite-shot counts and the projected remaining-mode branch state. The existing all-modes seeded sampling test still covers sampled full-mode counts.

File: piquasso/fermionic/fock/simulation_steps.py (hunk 2 above)
Comment (Rajesh270712) [reply piquasso/fermionic/fock/simulation_steps.py]: Simplified `_project_to_subspace` to iterate over the remaining-mode basis only, fill the measured modes with the sampled basis, and compute the original state-vector index directly. This removes the full-basis scan and the skip conditional.

**Review by Kolarovszki — APPROVED** [review]
This patch looks very good! Thank you so much for your contribution!

**Comment by Rajesh270712** [comment]
Thanks, I saw some build failure so resolved it . Lets merge the branch.

**Comment by Kolarovszki** [comment]
@Rajesh270712 can you comment on the original issue, so that I can assign you to it?

**Comment by Rajesh270712** [comment]
Done @Kolarovszki
